src/tls_openssl/future.rs

```diff
@@ -0,0 +1,174 @@
+//! Future types.
+
+use super::OpenSSLConfig;
+use pin_project_lite::pin_project;
+use std::io::{Error, ErrorKind};
+use std::time::Duration;
+use std::{
+    fmt,
+    future::Future,
+    io,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tokio::io::{AsyncRead, AsyncWrite};
+use tokio::time::{timeout, Timeout};
+
+use openssl::ssl::Error as OpenSSLError;
+use openssl::ssl::Ssl;
+use tokio_openssl::SslStream;
+
+// use tokio_rustls::{server::TlsStream, Accept, TlsAcceptor};
+
+pin_project! {
+    /// Future type for [`OpenSSLAcceptor`](crate::tls_openssl::OpenSSLAcceptor).
+    pub struct OpenSSLAcceptorFuture<F, I, S> {
+        #[pin]
+        inner: AcceptFuture<F, I, S>,
+        config: Option<OpenSSLConfig>,
+    }
+}
+
+impl<F, I, S> OpenSSLAcceptorFuture<F, I, S> {
+    pub(crate) fn new(future: F, config: OpenSSLConfig, handshake_timeout: Duration) -> Self {
+        let inner = AcceptFuture::InnerAccepting {
+            future,
+            handshake_timeout,
+        };
+        let config = Some(config);
+
+        Self { inner, config }
+    }
+}
+
+impl<F, I, S> fmt::Debug for OpenSSLAcceptorFuture<F, I, S> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("OpenSSLAcceptorFuture").finish()
+    }
+}
+
+pin_project! {
+    struct TlsAccept<I> {
+        #[pin]
+        tls_stream: Option<SslStream<I>>,
+    }
+}
+
+impl<I> Future for TlsAccept<I>
+where
+    I: AsyncRead + AsyncWrite + Unpin,
+{
+    type Output = io::Result<SslStream<I>>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let mut this = self.project();
+
+        match this
+            .tls_stream
+            .as_mut()
+            .as_pin_mut()
+            .map(|inner| inner.poll_accept(cx))
+            .expect("tlsaccept polled after ready")
+        {
+            Poll::Ready(Ok(())) => {
+                let tls_stream = this.tls_stream.take().expect("tls stream vanished?");
+
+                Poll::Ready(Ok(tls_stream))
+            }
+            Poll::Ready(Err(e)) => Poll::Ready(Err(Error::new(ErrorKind::Other, e))),
+            Poll::Pending => Poll::Pending,
+        }
+    }
+}
```

Comment:
This shouldn't be necessary. Doing it this way preserves same code structure as rustls version but I don't think extra complexity is worth it. What do you think?

Reply:
The problem is that the openssl library is different to rustls, where rustls is returning https://docs.rs/tokio-rustls/latest/tokio_rustls/struct.Accept.html from the call to accept(), and when polled yields the tlsstream. openssl doesn't do this, with https://docs.rs/tokio-openssl/latest/tokio_openssl/struct.SslStream.html#method.poll_accept returning `()` instead, so we need a way to "wrap" the the poll_accept so that we can combine that future with timeout. That's why I added this segment here so that we can properly do the poll with timeout. :)

Reply:
Manually selecting with sleep future can be done but I think you are right. It's just more complex.

Reply:
Yeah, that's why I opted this way, it's a bit clearer overall and keeps it in line with rustls even if more lines of code. 

I'm currently just adding the extra tests, docs, and cleanup and will submit the proper review shortly :)